plugins/grid-bookmark/src/index.ts

```diff
@@ -0,0 +1,134 @@
+        ),
+      })
+    })
+
+    pluginManager.extendPluggable('LinearGenomeView', LinearGenomeView => {
+      const { stateModel } = LinearGenomeView as ViewType
+      const newStateModel = stateModel.extend((self: LinearGenomeViewModel) => {
+        const superMenuItems = self.menuItems
+        const superRubberBandMenuItems = self.rubberBandMenuItems
+        return {
+          actions: {
+            activateBookmarkWidget() {
+              const session = getSession(self)
+              if (isSessionModelWithWidgets(session)) {
+                // @ts-ignore
+                let bookmarkWidget = session.widgets.get('GridBookmark')
+                if (!bookmarkWidget) {
+                  bookmarkWidget = session.addWidget(
+                    'GridBookmarkWidget',
+                    'GridBookmark',
+                    { view: self },
+                  )
+                }
+
+                session.showWidget(bookmarkWidget)
+                return bookmarkWidget
+              }
+
+              throw new Error('Could not open bookmark widget')
+            },
+
+            bookmarkCurrentRegion() {
+              const selectedRegions = self.getSelectedRegions(
+                self.leftOffset,
+                self.rightOffset,
+              )
+              const firstRegion = selectedRegions[0]
+              // @ts-ignore
+              const { widgets } = getSession(self)
+              let bookmarkWidget = widgets.get('GridBookmark')
+              if (!bookmarkWidget) {
+                this.activateBookmarkWidget()
+                bookmarkWidget = widgets.get('GridBookmark')
+              }
+              bookmarkWidget.addBookmark(firstRegion)
+            },
+          },
+          views: {
+            menuItems() {
+              const menuItems = superMenuItems()
+              menuItems.push(
+                { type: 'divider' },
+                {
+                  label: 'Open bookmark widget',
+                  icon: BookmarksIcon,
+                  // @ts-ignore
+                  onClick: self.activateBookmarkWidget,
+                },
+                {
+                  label: 'Bookmark current region',
+                  icon: BookmarkIcon,
+                  // @ts-ignore
+                  onClick: self.bookmarkCurrentRegion,
+                },
+              )
+              return menuItems
+            },
+
+            rubberBandMenuItems() {
+              const rubberBandMenuItems = superRubberBandMenuItems()
+              rubberBandMenuItems.push({
+                label: 'Bookmark region',
+                icon: BookmarkIcon,
+                onClick: () => {
+                  const { leftOffset, rightOffset } = self
+                  const selectedRegions = self.getSelectedRegions(
+                    leftOffset,
+                    rightOffset,
+                  )
+                  const firstRegion = selectedRegions[0]
+                  // @ts-ignore
```

Comment:
Probably needs a `isSessionModelWithWidgets` check like above to get rid of ts-ignore